R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+  df[,ids[1]] <- df$.ID
+  df$.ID <- NULL
+  return(df)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @param x a fitted \code{teeMod} model
+#' @keywords internal
+.prepare_design_matrix <- function(x, ...) {
+  design_obj <- x@Design
+  data <- x$call$data
+  name_clu <- colnames(design_obj@structure)[design_obj@column_index == "u"]
+  cid <- .merge_block_id_cols(data, name_clu)[, name_clu[1]]
+  
+  agg <- .aggregate_to_cluster(x)
+  bid <- agg[[1]][, agg[[2]]]  # block ids
+  
+  name_y <- as.character(x$terms[[2]]) # name of the outcome column
+  X1 <- model.matrix(x$call$offset@fitted_covariance_model) # design matrix
+  p <- ncol(X1)
+  
+  wc <- x$call$offset@fitted_covariance_model$weight
+  if (is.null(wc)) wc <- 1
+  X1 <- matrix(
+    rep(wc * (data[, name_y] - x$offset), p), ncol = p
+    ) * X1 # n by p, wic * residual * xi
+  
+  wi <- x$weights
+  X2 <- wi * x$residuals  # n by 1, wi[z] * (residual - rhoz) * z
+  
+  XX <- cbind(X1, X2)
+  XX <- aggregate(XX, by = list(cid), FUN = sum)[,-1]
+  
+  # multiple sqrt(product of #treated divided by block sizes) to XX by group
+  nbk <- design_table(design = design_obj, x = "treatment", y = "block")
+  const <- sqrt(nbk[, 1] * nbk[, 2] / rowSums(nbk))
+  XX <- sweep(XX, 1, const[bid], '*')
+  return(XX)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.cov_mat_est <- function(XXz, bidz){
+  cov0 <- tapply(1:nrow(XXz), bidz, function(s) cov(XXz[s,]))
+  covuu <- tapply(1:nrow(XXz), bidz, function(s) .add_vec(XXz[s,]))
+  covll <- tapply(1:nrow(XXz), bidz, function(s) .add_vec(XXz[s,], upper=FALSE))
+    
+  cov0u <- lapply(1:length(cov0), 
+                  function(s) if (is.na(cov0[[s]][1,1])) covuu[[s]] else cov0[[s]])
+  cov0l <- lapply(1:length(cov0), 
+                  function(s) if (is.na(cov0[[s]][1,1])) covll[[s]] else cov0[[s]])
+  
+  V00u <- Reduce('+', cov0u)
+  V00l <- Reduce('+', cov0l)
+  return(rbind(V00u, V00l))
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.add_mat_diag <- function(A, B){
+  d <- nrow(A)
+  A <- matrix(rep(diag(A), d), nrow = d, byrow = FALSE)
+  B <- matrix(rep(diag(B), d), nrow = d, byrow = TRUE)
+  #return(sqrt(abs(A * B)))
+  return((A + B) / 2)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.add_vec <- function(a, upper = TRUE){
+  if (nrow(a) > 1) return(0)
+  a <- as.numeric(a)
+  d <- length(a)
+  A <- matrix(rep(a, d), nrow = d, byrow = FALSE)
+  B <- matrix(rep(a, d), nrow = d, byrow = TRUE)
+  #if (upper) return(A*B + sqrt(abs(A*B)))
+  #else return(A*B - sqrt(abs(A*B)))
+  if (upper) return((A + B)^2 / 2)
+  else return(- (A - B)^2 / 2)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
+.cov01_est <- function(XX, zobs, bid){
```

Comment:
Same ask for documentation explaining what this function computes

Reply:
Added documentation. See commit https://github.com/benbhansen-stats/propertee/pull/185/commits/3f27a1e6c731038c3d2dd81cbee8451fdfe5bbcc